pyteal/ast/subroutine.py

```diff
@@ -389,15 +453,138 @@ def name(self) -> str:
+
+    This class is meant to be used as a function decorator. For example:
+
+        .. code-block:: python
+
+            @ABIReturnSubroutine
+            def abi_sum(toSum: abi.DynamicArray[abi.Uint64], *, output: abi.Uint64) -> Expr:
+                i = ScratchVar(TealType.uint64)
+                valueAtIndex = abi.Uint64()
+                return Seq(
+                    output.set(0),
+                    For(i.store(Int(0)), i.load() < toSum.length(), i.store(i.load() + Int(1))).Do(
+                        Seq(
+                            toSum[i.load()].store_into(valueAtIndex),
+                            output.set(output.get() + valueAtIndex.get()),
+                        )
+                    ),
+                )
+
+            program = Seq(
+                (to_sum_arr := abi.make(abi.DynamicArray[abi.Uint64])).decode(
+                    Txn.application_args[1]
+                ),
+                (res := abi.Uint64()).set(abi_sum(to_sum_arr)),
+                abi.MethodReturn(res),
+                Int(1),
+            )
+    """
+
+    def __init__(
+        self,
+        fn_implementation: Callable[..., Expr],
+    ) -> None:
+        self.output_kwarg_info: Optional[
+            OutputKwArgInfo
+        ] = self._output_name_type_from_fn(fn_implementation)
+
+        internal_subroutine_ret_type = TealType.none
+        if self.output_kwarg_info:
+            internal_subroutine_ret_type = (
+                self.output_kwarg_info.abi_type.storage_type()
+            )
+
+        output_kwarg_name = None
+        if self.output_kwarg_info:
+            output_kwarg_name = self.output_kwarg_info.name
+
+        # output ABI type is void, return_type = TealType.none
+        # otherwise, return_type = ABI value's storage_type()
+        self.subroutine = SubroutineDefinition(
+            fn_implementation,
+            return_type=internal_subroutine_ret_type,
+            abi_output_arg_name=output_kwarg_name,
+        )
+
+    @staticmethod
+    def _output_name_type_from_fn(
+        fn_implementation: Callable[..., Expr]
+    ) -> Optional[OutputKwArgInfo]:
+        if not callable(fn_implementation):
+            raise TealInputError("Input to ABIReturnSubroutine is not callable")
+        sig = signature(fn_implementation)
+        fn_annotations = get_annotations(fn_implementation)
+
+        potential_abi_arg_names = [
+            k for k, v in sig.parameters.items() if v.kind == Parameter.KEYWORD_ONLY
+        ]
+
+        match potential_abi_arg_names:
+            case []:
+                return None
+            case [name]:
+                annotation = fn_annotations.get(name, None)
+                if annotation is None:
+                    raise TealInputError(
+                        f"ABI subroutine output-kwarg {name} must specify ABI type"
+                    )
+                type_spec = abi.type_spec_from_annotation(annotation)
+                return OutputKwArgInfo(name, type_spec)
```

Comment:
I originally thought we'd force all functions to use the "output" keyword argument, but I see this implementation allows you to pick any name. Did you have uses cases or benefits in mind when you made this decision? I'm trying to evaluate whether it's better or worse than fixing the argument name.

Reply:
Ok, I see. My current thoughts are that we should be more conservative initially and limit the name to "output" or something similar that makes it clear the argument is meant to capture the return value.

Then going forward we can always drop this restriction without jeopardizing existing code.

Reply:
ok I see. I added a few lines of restrictions on naming for kwarg in `ABIReturnSubroutine`.

LMK if it looks right